Rewrite to use argparse

ocharles opened the pull request:
This commit mostly rewrites nginx-upstream-manager to use argparse
instead of optparse. This gives us the ability to have more centralised
command line parsing, better usages for free, and stricter validation.

I have also changed a few defaults from '**file** + path' to use
either command line overrides, or be relative to the config file itself.
This is because I would like nginx-upstream-manager to be installed,
rather than a script that just lies around.

ianmcorvidae commented:
In general I'm not happy with using argparse, because it's not supported until python 2.7, as I already told you before.

ianmcorvidae commented on upstream_manager.py:
```diff
@@ -45,7 +45,6 @@ def process(self):
                 config[cluster_key][server_key].update(config_raw[cluster_key][server_key])
 
             config[cluster_key]['_ip_hash'] = bool(config_raw[cluster_key].get('_ip_hash', False))
-            config[cluster_key]['_file'] = dirname(__file__) + '/' + config_raw[cluster_key].get('_file', cluster_key + '.conf')
```
Not sure why this has gone away; it configures the file to save _to_.

ocharles replied:
It can still be used, but it just doesn't override what is in the configuration. Instead, the configuration file should contain the absolute path.

ianmcorvidae replied:
This (with the line below) is now better, but it still means there's no default configuration. I'd prefer if this had some sort of sensible default, such as the previous system where it created a file in the same directory as config.yaml, corresponding to the cluster name. That might, in fact, be a reasonable choice here as well. If not, it needs error handling; with the present code it'll kill the whole script with KeyError if there's no _file configured.

ocharles replied:
It doesn't die with a KeyError, because the main entry point checks that either `--output` is present, or `cluster.filename` is set.

ianmcorvidae commented on upstream_manager.py:
```diff
@@ -109,7 +108,7 @@ def __init__(self, name, config):
         self.servers = [Server(name, config[name]) for name in config.keys() if name[0] != '_']
 
-    def save(self, **kwargs):
+    def render(self, **kwargs):
```
Unclear what this change is meant to accomplish; I don't really oppose it but I do prefer the notion that .save() writes the file associated with the object you have, whether it's a Config or a Cluster.

ocharles replied:
I went for `render` because unlike `Config` you don't load these from the same file you save to. Rather, you load from an abstract configuration, and then render this down to an nginx configuration.

ianmcorvidae commented on upstream_manager.py:
```diff
@@ -196,73 +195,86 @@ def rotate_action(config, cluster, *other_args):
         config.disable(cluster, to_disable)
     config.save()
-    config.cluster(cluster.name).save()
+    cluster.render()
```
This won't work correctly, because you need to reload the cluster after you've rewritten the config. That's why it used the other syntax.

ianmcorvidae commented on upstream_manager.py:
```diff
@@ -196,73 +195,86 @@ def rotate_action(config, cluster, *other_args):
         config.enable(cluster, to_enable)
     config.save()
-    config.cluster(cluster.name).save()
+    cluster.render()
```
(likewise)

ianmcorvidae commented on upstream_manager.py:
```diff
@@ -196,73 +195,86 @@ def rotate_action(config, cluster, *other_args):
     config.save()
-    config.cluster(cluster.name).save()
-    print "Changed %s weight to %s" % (server, value)
+    cluster.render()
```
(likewise)

ianmcorvidae commented on upstream_manager.py:
```diff
@@ -196,73 +195,86 @@ def rotate_action(config, cluster, *other_args):
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-c', '--config', type=config,
+                        help='Path to the abstract YAML configuration',
+                        default='config.yaml')
```
If I read this correctly it forces running this from a specific directory if you intend to take advantage of the default, avoiding of which was the whole point of the **file** usage in the first place

ocharles replied:
That is correct. Previously you relied on the script to be in the same directory as the configuration itself, but I'd like this script to be packaged in more of a standard way - installing to /usr/bin, and having configuration files in /etc. Now you cd to /etc and run it from there, or just run it with an explicit --config.

ianmcorvidae commented on upstream_manager.py:
```diff
@@ -196,73 +195,86 @@ def rotate_action(config, cluster, *other_args):
+    parser.add_argument('-c', '--config', type=config,
+                        help='Path to the abstract YAML configuration',
+                        default='config.yaml')
+    parser.add_argument('-o', '--output', help='path to generate nginx configuration')
```
This should not need to be passed every time. As I mentioned above, this should be entirely from config.yaml.

ocharles replied:
It doesn't have to be passed, but you can override what is in the configuration file. I found this useful testing the script.

ianmcorvidae commented on upstream_manager.py:
```diff
@@ -196,73 +195,86 @@ def rotate_action(config, cluster, *other_args):
+                                     help='generate a concrete nginx configuration from the abstract cluster description')
+    generate.set_defaults(run=generate_action)
+
+    # The 'rotate' command does... something
```
The rotate command is what we use for musicbrainz-server deployments. The first time you call it, it disables one server and prints it out, which fabric should capture and deploy to that server. Then, when called again, it brings that one back in and takes out the next, printing out the IP; then you can deploy to that, and so forth until it's done all non-disabled/non-'down' servers, at which point it'll print out 'Done' and you know you're finished.

ocharles replied:
Oh, this wasn't meant to sound so sarcy! I was working through them one by
one and commenting as I went but I forgot to fix this comment. It does
indeed do what you said when I ran and tested that command :) this was just
a place holder for me to expand the "something"
On 5 Jul 2013 20:00, "Ian McEwen" notifications@github.com wrote:

> In upstream_manager.py:
> 
> > -    actions[action](config, cluster, args[2:], options, parser)
> > -    parser = argparse.ArgumentParser()
> > -    parser.add_argument('-c', '--config', type=config,
> > -                        help='Path to the abstract YAML configuration',
> > -                        default='config.yaml')
> > -    parser.add_argument('-o', '--output', help='path to generate nginx configuration')
> > -    parser.add_argument('cluster')
> >   +
> > -    subparsers = parser.add_subparsers()
> >   +
> > -    # The 'generate' command generates nginx configuration files
> > -    generate = subparsers.add_parser('generate',
> > -                                     help='generate a concrete nginx configuration from the abstract cluster description')
> > -    generate.set_defaults(run=generate_action)
> >   +
> > -    # The 'rotate' command does... something
> 
> The rotate command is what we use for musicbrainz-server deployments. The
> first time you call it, it disables one server and prints it out, which
> fabric should capture and deploy to that server. Then, when called again,
> it brings that one back in and takes out the next, printing out the IP;
> then you can deploy to that, and so forth until it's done all
> non-disabled/non-'down' servers, at which point it'll print out 'Done' and
> you know you're finished.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/ianmcorvidae/nginx-upstream-manager/pull/2/files#r5048309
> .

ianmcorvidae replied:
Ah, okay then. Well, still expand it out before we ship this!

ocharles commented:
argparse works fine on lenny via apt-get. I'll respond to your other
comments when I'm at home though, but I don't think argparse is a show
stopper

ianmcorvidae commented:
Ah, I see, it's also on PyPI. Make sure to add it to requirements.txt, then, if nothing else for tests (once they actually test calling the script itself, which I'd given up on wrangling travisci to do -- may move this over to ci.mb anyway though)

ocharles commented:
I think I've addressed your comments. The only one that hasn't been addressed is:

> This won't work correctly, because you need to reload the cluster after you've rewritten the config. That's why it used the other syntax.

With the code how it is at the moment all the commands do appear to work correctly, so could you expand on what it is that I've broken?

ianmcorvidae commented:
Right now you have it save the config then immediately do cluster.render. However, that won't/shouldn't work, because e.g. config.disable(cluster, to_disable) changes the config object but _not_ the cluster object. So it'll save the changes properly to the config.yaml, but will save the pre-change version to the generated nginx config file. The previous code had config.cluster(cluster.name).save() instead, because that pulls the changed version from the config object rather than using the version extracted from the configuration earlier in the same invocation of the script.

ocharles commented:
Ok, got it. I've pushed a new version out that should address that be reverting back to what you had.

ianmcorvidae commented on upstream_manager.py:
```diff
@@ -212,10 +210,8 @@ def disable_action(config, cluster, args, options, parser):
     config.cluster(cluster.name).save()
```
Now that you turned this back it's wrong again, because you changed 'save' to 'render'.

ocharles replied:
Argh! Good catch. I'll try again and actually test it this time >.<

ocharles replied:
OK, fixed and I tested `enable`, `disable`, `generate`, `weight` and `rotate`. All did what I expected.

ianmcorvidae commented:
I guess I hadn't looked at this since your last push; this looks ready to go, I think.
